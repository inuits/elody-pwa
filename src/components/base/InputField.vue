<template>
  <label
    class="block my-2"
    :class="type === 'checkbox' ? 'justify-end flex flex-row-reverse' : ''"
  >
    <span v-if="label" class="ml-1 text-neutral-700 text-sm">{{ label }}</span>
    <div :class="[{ 'mr-4': icon }, inputContainerStyle]">
      <unicon
        v-if="icon"
        :name="icon"
        :class="[`h-full pl-2 text-neutral-700 bg-${bgColor}`]"
      />
      <input
        v-model="inputValue"
        :disabled="isDisabled"
        v-bind="$attrs"
        :class="[`bg-${bgColor}`, inputStyle]"
        :type="type"
        name=""
      />
    </div>
  </label>
</template>

<script lang="ts">
import { computed, defineComponent } from "vue";
import type { PropType } from "vue";
import { debounce } from "ts-debounce";
export const lableStyle = 'ml-1 text-neutral-700 text-sm"';
export const inputContainerStyle =
  "flex flex-row bg-neutral-20 border border-neutral-50 rounded";
export const inputStyle =
  " py-2 pl-4  w-full rounded min-w-48 text-neutral-700 text-sm focus:outline-none";

export default defineComponent({
  name: "InputField",
  inheritAttrs: false,
  props: {
    label: { type: String, default: "" },
    modelValue: {
      type: String,
      default: "",
    },
    debounce: { type: Boolean, default: false },
    debounceWait: { type: Number, default: 400 },
    icon: { type: String, default: undefined },
    bgColor: { type: String, default: "neutral-0" },
    name: { type: String, default: "", required: false },
    isDisabled: { type: Boolean, default: false, required: false },
    type: { type: String, required: false, default: "text" },
  },
  emits: ["update:modelValue"],
  setup(props, { emit }) {
    let emitValue = (value: string) => emit("update:modelValue", value);

    if (props.debounce) {
      emitValue = debounce(emitValue, props.debounceWait);
    }

    const inputValue = computed<any>({
      get() {
        return props.modelValue;
      },
      set(value: string | boolean) {
        emitValue(value.toString());
      },
    });
    return { inputValue, inputStyle, inputContainerStyle };
  },
});
</script>
