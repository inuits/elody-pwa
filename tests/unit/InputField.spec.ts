import { shallowMount } from '@vue/test-utils';
import InputField from '../../src/components/base/InputField.vue';
jest.mock('@/types');

let wrapper: any;

describe('InputField.vue', () => {
  wrapper = shallowMount(InputField, {
    props: {
      placeholder: 'Search Asset Library...',
      label: 'search',
      search: 'asset',
      debounce: false,
    },
  });

  it('Renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the correct label', () => {
    const label = wrapper.find('label');
    expect(label.text()).toMatch('');
  });

  it('Updated ref inputValue ', () => {
    const input = wrapper.find('input');
    input.setValue('dummy');
    wrapper.vm.$nextTick(() => {
      expect(wrapper.inputValue.value).toBe('dummy');
      expect(wrapper.emitted('update:modelValue')).toEqual([['dummy']]);
    });
  });
});
