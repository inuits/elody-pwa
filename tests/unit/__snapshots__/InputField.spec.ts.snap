// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField.vue Renders correctly 1`] = `
<label
  class="block my-2"
>
  <span
    class="ml-1 text-neutral-700 text-sm"
  >
    search
  </span>
  <div
    class="flex flex-row bg-neutral-20 border border-neutral-30 rounded mr-4"
  >
    <unicon
      class="h-full pl-2 text-neutral-700"
      name="search-glass"
    />
    <input
      class="py-2
          pl-4
          w-48
          min-w-0
          bg-neutral-20
          text-neutral-700 text-sm
          focus:outline-none"
      placeholder="Search Asset Library..."
      search="asset"
      type="text"
    />
  </div>
</label>
`;
